### DISCUSSION TITLE
Project Structure -Best Strategy?

### DISCUSSION THREAD

**Discussion by abcnow** [serverpod/serverpod#869] [General]
Hi There: I was looking at a few of your projects and documentations and it I couldn't find anything related to how to structure a project with serverpod.. I even opened the code for your serverpod_chat module (https://github.com/serverpod/serverpod/tree/main/modules/serverpod_chat/serverpod_chat_flutter/lib/src) and everything there is together.. views with widgets and services, etc -everything is under your lib file.. let's say that you are trying to do a serious chat application with 4 main pages:
- Login
- Registration
- 1-on-1 chat
- Group Chat

If I have to structure that with flutter and nodejs.. you will just create something like this
- FrontEnd 
   ->Assets
   ->Constants (error_handling, global variables, etc)
   ->Common (custom wigets, textfields, etc)
   ->models (user, etc)
   -> Features (Auth, ChatPageHome, etc)
   --Router.dart
   --Main.dart

-BackEnd (Nodejs)
    ->Middlewares
    ->Models
    ->Routes
   --Index

----What would be the best approach with Serverpod..? I am asking this because in your documents you are recommending to work with Modules.. and I would like to hear what's the best strategy to follow... etc. Thanks in advance!

**Comment by yahu1031** [comment]
@abcnow The flutter code can be restructured as your wish. But client code should not be touched other than for required changes. Now slowly serverpod started adapting to structuring the backend(server) code as the developer wishes(not completely I see).

**Reply by abcnow** [reply]
> @abcnow The flutter code can be restructured as your wish. But client code should not be touched other than for required changes. Now slowly serverpod started adapting to structuring the backend(server) code as the developer wishes(not completely I see).

thanks a lot for your fast response.. can you elaborate a little more on "started adapting to structuring.."? what's the plan on that and why?

**Reply by yahu1031** [reply]
@abcnow This was never meant by Viktor's brain 🤣 But as we have seen before(early stage) we don't have subfolders for the server API endpoints. Now it supports having multiple subfolders in server APIs/Models. But surely(may be) in the future you can write your code as you like in server folders too and that will generate a systematic file structure in client code.

**Comment by vlidholt** [comment]
It's up to you how you want to structure your Flutter app, there are many options and preferences there.

For the server side. Typically, I would place light weight API calls in the `endpoints` directory. Then you can create a business layer where you keep the logic of your backend. This can also be smart in case you at some point want to share that code with the web server API.

The modules are used if you want to write reusable code that you want to share between different apps or make public. Similar to Dart packages.

**Reply by abcnow** [reply]
@vlidholt can you elaborate a little more on this for example with a todo app?

---

For the server side. Typically, I would place light weight API calls in the endpoints directory. Then you can create a business layer where you keep the logic of your backend. This can also be smart in case you at some point want to share that code with the web server API.?
